### PR TITLE
OSDOCS#6990: Add CPMS variable root volume types

### PR DESCRIPTION
**Pull request by snarayan-redhat**
Version(s): 4.14

Issue: https://issues.redhat.com/browse/OSDOCS-6990

Link to docs preview: https://63861--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-custom#installation-configuration-parameters-additional-osp_installing-openstack-installer-custom

QE review:
- [x] QE has approved this change. @itzikb-redhat
- [x] SME has approved this change. @EmilienM / @pierreprinetti

### CODE REVIEW

File: modules/installation-configuration-parameters.adoc [hunk 1]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
 |`controlPlane.platform.openstack.rootVolume.size`
 |For control plane machines, the size in gigabytes of the root volume. If you do not set this value, machines use ephemeral storage.
 |Integer, for example `30`.
 
+|`controlPlane.platform.openstack.rootVolume.types`
+|For control plane machines, the root volume's types.
+|A list of strings, for example, {`performance`, `standard`}. ^[1]^
+
 |`controlPlane.platform.openstack.rootVolume.type`
 |For control plane machines, the root volume's type.
```
Comment (EmilienM) [inline modules/installation-configuration-parameters.adoc:955]: Other deprecated strings have something like:

```
This property is deprecated. (...)
```

I would add it too, to make it clear, even if you have an annotation.

**Review by EmilienM — CHANGES_REQUESTED** [review]
LGTM overall, but I have a minor change request, if you think that makes sense.

**Comment by EmilienM** [comment]
cc @pierreprinetti just FYI

File: modules/installation-configuration-parameters.adoc [hunk 2]
```diff
@@ -940,15 +940,15 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
+|For compute machines, the root volume's type. This property is deprecated.
 |String, for example, `performance`. ^[2]^
 
 |`controlPlane.platform.openstack.rootVolume.size`
 |For control plane machines, the size in gigabytes of the root volume. If you do not set this value, machines use ephemeral storage.
 |Integer, for example `30`.
 
 |`controlPlane.platform.openstack.rootVolume.types`
-|For control plane machines, the root volume's types.
+|For control plane machines, the root volume's types. This property is deprecated.
```
Comment (EmilienM) [inline modules/installation-configuration-parameters.adoc:951]: Wrong line :D

Comment (snarayan-redhat) [reply modules/installation-configuration-parameters.adoc]: oops. Fixed.

File: modules/installation-configuration-parameters.adoc [hunk 3]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
 |Integer, for example `30`.
 
+|`compute.platform.openstack.rootVolume.types`
+|For compute machines, the root volume's types.
+|A list of strings, for example, {`performance`, `standard`}. ^[1]^
+
 |`compute.platform.openstack.rootVolume.type`
-|For compute machines, the root volume's type.
-|String, for example `performance`.
+|For compute machines, the root volume's type. This property is deprecated.
```
Comment (EmilienM) [inline modules/installation-configuration-parameters.adoc:943]:
```suggestion
|For compute machines, the root volume's type. This property is deprecated and is replaced by `compute.platform.openstack.rootVolume.types`.
```

File: modules/installation-configuration-parameters.adoc [hunk 4]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
 |Integer, for example `30`.
 
+|`controlPlane.platform.openstack.rootVolume.types`
+|For control plane machines, the root volume's types.
+|A list of strings, for example, {`performance`, `standard`}. ^[1]^
+
 |`controlPlane.platform.openstack.rootVolume.type`
-|For control plane machines, the root volume's type.
-|String, for example `performance`.
+|For control plane machines, the root volume's type. This property is deprecated.
```
Comment (EmilienM) [inline modules/installation-configuration-parameters.adoc:955]:
```suggestion
|For control plane machines, the root volume's type. This property is deprecated and is replaced by `controlPlane.platform.openstack.rootVolume.types`.
```

File: modules/installation-configuration-parameters.adoc [hunk 5]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
+|For compute machines, the root volume's type. This property is deprecated.
+|String, for example, `performance`. ^[2]^
 
 |`controlPlane.platform.openstack.rootVolume.size`
 |For control plane machines, the size in gigabytes of the root volume. If you do not set this value, machines use ephemeral storage.
 |Integer, for example `30`.
 
+|`controlPlane.platform.openstack.rootVolume.types`
+|For control plane machines, the root volume's types.
+|A list of strings, for example, {`performance`, `standard`}. ^[1]^
```
Comment (pierreprinetti) [inline modules/installation-configuration-parameters.adoc:952]: While not incorrect, I find the example misleading. Master machines must be equal in resources and performance; using "standard" and "performance" may imply differently. Also, masters are typically exactly three. Please consider:

```suggestion
|A list of strings, for example, {`performance-host1`, `performance-host2`, `performance-host3`}. ^[1]^
```

File: modules/installation-configuration-parameters.adoc [hunk 6]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
-|String, for example `performance`.
+|For compute machines, the root volume's type. This property is deprecated.
+|String, for example, `performance`. ^[2]^
 
 |`controlPlane.platform.openstack.rootVolume.size`
 |For control plane machines, the size in gigabytes of the root volume. If you do not set this value, machines use ephemeral storage.
 |Integer, for example `30`.
 
+|`controlPlane.platform.openstack.rootVolume.types`
+|For control plane machines, the root volume's types.
```
Comment (pierreprinetti) [inline modules/installation-configuration-parameters.adoc:951]: Each Control plane node has its own root volume, each volume its type. Please consider:

```suggestion
|For control plane machines, the root volume types.
```

File: modules/installation-configuration-parameters.adoc [hunk 7]
```diff
@@ -964,6 +972,11 @@ This property is deprecated. To use a flavor as the default for all machine pool
 |String, for example `m1.xlarge`.
 |====
 
+. If the machine pool defines `zones`, the count of types can either be a single item or match the number of items in `zones`. For example, the count of types cannot be 2 if there are 3 items in `zones`.
+
+. Because this property is deprecated, the installer converts the provided value into a list of string with one item, and populates the `types` field with this list item.
```
Comment (pierreprinetti) [inline modules/installation-configuration-parameters.adoc:977]: This note should refer to the `type` field, but I see no reference to it. Is it more evident when the document is formatted?

Comment (snarayan-redhat) [reply modules/installation-configuration-parameters.adoc:977]: You can check the preview link to see how it renders. It is added as a footnote with reference.

This mechanism is explained to let users know that `type` is still supported and converted to `types`. This also helps them know that they need not redo any work if they have any existing references to `type`. 
Do you think it needs to be said differently? Like `Existing references to this field need not be changed` ??

Comment (pierreprinetti) [reply modules/installation-configuration-parameters.adoc:977]: >You can check the preview link to see how it renders. It is added as a footnote with reference.

Ok, checked! Looks perfect.

>This also helps them know

That makes sense, I just hope we're not going too much into the weeds. But that's your call, not mine :)

Comment (snarayan-redhat) [reply modules/installation-configuration-parameters.adoc:977]: Sure. I will reword this to touch upon the conversion and just let users know that the values under `type` is converted to `types` (without the single list item explanation)

Comment (snarayan-redhat) [reply modules/installation-configuration-parameters.adoc]: @pierreprinetti I have reworded the current statement. Can you have a look and let me know it's ok?

I have made all the suggested changes. Let me know if there are any more suggestions. If not, can you give an lgtm?

**Review by pierreprinetti — APPROVED** [review]
LGTM! Thank you

File: modules/installation-configuration-parameters.adoc [hunk 8]
```diff
@@ -935,17 +935,25 @@ Additional {rh-openstack} configuration parameters are described in the followin
...
 |Integer, for example `30`.
 
+|`compute.platform.openstack.rootVolume.types`
+|For compute machines, the root volume types.
+|A list of strings, for example, {`performance-host1`, `performance-host2`, `performance-host3`}. ^[1]^
+
 |`compute.platform.openstack.rootVolume.type`
-|For compute machines, the root volume's type.
-|String, for example `performance`.
+|For compute machines, the root volume's type. This property is deprecated and is replaced by `compute.platform.openstack.rootVolume.types`..
```
Comment (EmilienM) [inline modules/installation-configuration-parameters.adoc:943]: why `..` at the end?

**Comment by itzikb-redhat** [comment]
I think we still need to document things like troubleshooting rootVolumes etc but for these specific parameters LGTM

**Comment by EmilienM** [comment]
@itzikb-redhat 

> I think we still need to document things like troubleshooting rootVolumes

Could you please be more specific so @snarayan-redhat can add some content if needed?

**Comment by itzikb-redhat** [comment]
> @itzikb-redhat
> 
> > I think we still need to document things like troubleshooting rootVolumes
> 
> Could you please be more specific so @snarayan-redhat can add some content if needed?
Now that I think about it we already have a ticket asking for CPMS troubleshooting so we'll probably have content there but it's important to check for CPMS existence after installation and check the logs of the CCPMSO after installation/upgrade since there is no indication from the installation/update itself (control-plane-machine-set-operator pod in the openshift-machine-api namespace)

**Comment by itzikb-redhat** [comment]
@EmilienM Shouldn't we add that compute.platform.openstack.rootVolume.zones  are Cinder AZs and are mandatory when compute.platform.openstack.rootVolume.types and compute.platform.openstack.zones are defined?

**Comment by EmilienM** [comment]
> @EmilienM Shouldn't we add that compute.platform.openstack.rootVolume.zones are Cinder AZs and are mandatory when compute.platform.openstack.rootVolume.types and compute.platform.openstack.zones are defined?

I agree but I think this might needs another bug report.

@snarayan-redhat FYI https://github.com/openshift/installer/commit/f31a29b738f96b734d6301dbd873950abe8a91b6#diff-8d13bb62fd33c601f7f64e3a9c861445f547962a0079d20613daab0c07dd73b8 if that helps.

**Comment by snarayan-redhat** [comment]
> @EmilienM Shouldn't we add that compute.platform.openstack.rootVolume.zones are Cinder AZs and are mandatory when compute.platform.openstack.rootVolume.types and compute.platform.openstack.zones are defined?

I agree with @EmilienM . This will need a separate ticket. Current ticket focuses only on root volume types. 

> > @itzikb-redhat
> > > I think we still need to document things like troubleshooting rootVolumes
> > 
> > 
> > Could you please be more specific so @snarayan-redhat can add some content if needed?

> > Now that I think about it we already have a ticket asking for CPMS troubleshooting so we'll probably have content there but it's important to check for CPMS existence after installation and check the logs of the CCPMSO after installation/upgrade since there is no indication from the installation/update itself (control-plane-machine-set-operator pod in the openshift-machine-api namespace)

Even this could either go with the troubleshooting ticket or can open a separate ticket.
